### ISSUE TITLE
Federation Project: Authentication Protocol

### ISSUE THREAD

**Issue by 1florentin** [dedis/popstellar#1800]
## Description
As part of the federation sub-project we want to implement an authentication protocol to "link" two LAOs together. For this purpose we designed the following schematics and messages.

This issue is meant to serve as a central place to reference in subsequent pull requests (implementing the protocol) and for discussion.

### Schematic Illustration
<img align="center" width="800" src="https://github.com/dedis/popstellar/assets/20774003/906531fa-8f40-4b2e-891c-c0dcd4745083" alt="schematic">

### Sequential Illustration
<img align="left" src="https://github.com/dedis/popstellar/assets/20774003/28b3eb37-4d6c-43d7-9d8b-d9c5a6afd8ad"  width="500" alt="schematic">

The two organizers meet in person to exchange the *conn_details* message.

They also decide on a time slot to perform their next roll-calls---these have to happen simultaneously.
<br clear="left"/>
## TODOs (for this issue)
* Add message explainations
* Add explanation regarding trust boundaries between the two connected LAOs.

**Comment by arnauds5** [comment]
With this protocol design, the exchange of conn_details happens only with QR-code, without any new message, is it correct ?
Maybe changing the name from conn_init/expect/etc to something like federation_init, could avoid any confusion with the basic server to server connection ?

The conn_init/conn_expect messages schemas would be something like the following ?

conn_init
```json
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$id": "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/query/method/message/data/connInit.json",
    "description": "Sent by an organizer client to its server, initiates a connection to a remote LAO",
    "type": "object",
    "properties": {
        "object": {
            "const": "conn"
        },
        "action": {
            "const": "init"
        },
        "lao_id": {
            "type": "string",
            "contentEncoding": "base64",
            "$comment": "ID of the remote LAO"
        },
        "server_address": {
            "type": "string",
            "$comment": "public address of the remote organizer server"
        },
        "public_key": {
            "type": "string",
            "contentEncoding": "base64",
            "$comment": "public key of the remote organizer"
        },
        "challenge": {
            "type": "object",
            "properties": {
                "value": {
                    "type": "integer",
                    "minimum": 0
                },
                "valid_until": {
                    "type": "integer",
                    "description": "[Timestamp] of the expiration time",
                    "minimum": 0
                }
            },
            "additionalProperties": false,
            "required": ["value", "valid_until"]
        }
    },
    "additionalProperties": false,
    "required": [
        "object",
        "action",
        "lao_id",
        "server_address",
        "public_key",
        "challenge"
    ]
}
```
example of conn_init
```json
{
    "object": "conn",
    "action": "init",
    "lao_id": "fzJSZjKf-2cbXH7kds9H8NORuuFIRLkevJlN7qQemjo=",
    "public_key": "J9fBzJV70Jk5c-i3277Uq4CmeL4t53WDfUghaK0HpeM=",
    "server_address": "wss://epfl.ch:9000/server",
    "challenge": {
        "value": 1234567890,
        "valid_until": 1712854874
    }
}
```

conn_expect
```json
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$id": "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/query/method/message/data/connExpect.json",
    "description": "Sent by an organizer client to its server, signals that a connection from a remote LAO is expected",
    "type": "object",
    "properties": {
        "object": {
            "const": "conn"
        },
        "action": {
            "const": "expect"
        },
        "lao_id": {
            "type": "string",
            "contentEncoding": "base64",
            "$comment": "ID of the remote LAO"
        },
        "server_address": {
            "type": "string",
            "$comment": "public address of the remote organizer server"
        },
        "public_key": {
            "type": "string",
            "contentEncoding": "base64",
            "$comment": "public key of the remote organizer"
        },
        "challenge": {
            "type": "object",
            "properties": {
                "value": {
                    "type": "integer",
                    "minimum": 0
                },
                "valid_until": {
                    "type": "integer",
                    "description": "[Timestamp] of the expiration time",
                    "minimum": 0
                }
            },
            "additionalProperties": false,
            "required": ["value", "valid_until"]
        }
    },
    "additionalProperties": false,
    "required": [
        "object",
        "action",
        "lao_id",
        "server_address",
        "public_key",
        "challenge"
    ]
}
```

example of conn_expect
```json
{
    "object": "conn",
    "action": "expect",
    "lao_id": "lsWUv1bKBQ0t1DqWZTFwb0nhLsP_EtfGoXHny4hsrwA=",
    "public_key": "UvViTxoKsB3XVP_ctkmOKCJpMWb7fCzrcb1XDmhNe7Q=",
    "server_address": "wss://ethz.ch:9000/server",
    "challenge": {
        "value": 1234567890,
        "valid_until": 1712854874
    }
}
```

**Comment by ljankoschek** [comment]
The only thing I would notice is that maybe we could add a pattern to the server_address like:
"pattern": "^(ws|wss):\\/\\/.*(:\\d{0,5})?\\/.*$"

Otherwise it looks good to me!

**Comment by 1florentin** [comment]
> With this protocol design, the exchange of conn_details happens only with QR-code, without any new message, is it correct ? 

Yes, it's exchanged via QR code (and only between frontends), but still a structure for the data to be exchanged would be needed, or?

> Maybe changing the name from conn_init/expect/etc to something like federation_init, could avoid any confusion with the basic server to server connection ?

Probably a good idea!

**Comment by 1florentin** [comment]
One thing to consider is, that in the *challenge* message/object, the value field is currently of type Number (this was just a first abstract choice).

It would probably be good for this value to be fixed amount (i.e. 32) random bytes (sourced from a secure (P)RNG like crypto/rand in Go) represented as hex or base64 or similar.

What do you think? @arnauds5 @ljankoschek

**Comment by onsriahi14** [comment]
> > One thing to consider is, that in the _challenge_ message/object, the value field is currently of type Number (this was just a first abstract choice).
> > It would probably be good for this value to be a fixed amount (i.e. 32) random bytes (sourced from a secure (P)RNG like crypto/rand in Go) represented as hex or base64 or similar.
> > What do you think? @arnauds5 @ljankoschek
> 
> yeah I think you are right that would be a better choice than a Number.
> 
> Also yes I think you are right @arnauds5 a "federation_[name}" would be nicer!
I'd go for the fixed amount of random bytes for Scala as well! and Yes it's better to change the name like Arnaud said.

**Comment by arnauds5** [comment]
> One thing to consider is, that in the _challenge_ message/object, the value field is currently of type Number (this was just a first abstract choice).
> 
> It would probably be good for this value to be a fixed amount (i.e. 32) random bytes (sourced from a secure (P)RNG like crypto/rand in Go) represented as hex or base64 or similar.
> 
@1florentin 
Yes, a fixed amount of random bytes would probably be better than a number/integer.
Depending on the number or bytes, we should check that the QR-code version we use, has enough data size.

**Comment by quadcopterman** [comment]
Another question would be : how do we precisely encode the connection data into a QR code. Do we just encode to JSON, get the corresponding string and create a QR code from it ?

**Comment by 1florentin** [comment]
> Another question would be : how do we precisely encode the connection data into a QR code. Do we just encode to JSON, get the corresponding string and create a QR code from it ?

IIRC, there are already QR codes being generated in the frontend implementations (e.g. rollcall QR codes)---can we not just look at how that's done there and do the same? (disclaimer: im not on any frontend team)

**Comment by arnauds5** [comment]
Some questions:

1.a) Is the challenge generated by S2 after a federation_expect ?
1.b) If yes, then the federation_expect should not have any challenge field ?

2.a) Is it possible for S1 to only send the challenge signed by Alice secret key ? (S1 only knowns the secret key S1, not Alice secret key)
2.b) If not, we could maybe forward the federation_init message instead of the challenge, that message will already be signed by Alice when S1 received it. Are we losing some security property if done that way ?
2.c) Is it possible for S2 to send a federation_result signed by bob secret key, without having the key ? Or is it signed using the secret key of the server S2 instead ?

**Comment by 1florentin** [comment]
> 1.a) Is the challenge generated by S2 after a federation_expect ? 1.b) If yes, then the federation_expect should not have any challenge field ?

No, the challenge is first retrieved by the organizer from the server. There has to be another message to do this.

So I propose we define a *challenge_request* message that is being sent from the organizer to the server. If we want to continue with the Alice and Bob example:

<pre>
Bob                            S2
     ---challenge_request-->
                .
                .
                .
</pre>

The protocol parts following this message are already defined---see the above specification and sequence diagram.

The messages exchanged between from the organizers to the servers are of course always signed by the organizer---I believe this is already implemented as messages to LAOs from clients already have to be signed (?).

> 2.a) Is it possible for S1 to only send the challenge signed by Alice secret key ? (S1 only knowns the secret key S1, not Alice secret key)

We discussed this in one of our last meetings, the idea was that Alice also signs the message object inside of the *federation_init* message. 

> 2.c) Is it possible for S2 to send a federation_result signed by bob secret key, without having the key ? Or is it signed using the secret key of the server S2 instead ?

Yes, S2 does not have the secret key of Bob. But the message **does have to be signed by Bob** or else we violate the trust assumptions we defined. (Remember, servers of the opposite LAO are not trusted. Any client and any server in any LAO only ever has to trust their own organizer)

This does make it difficult to sign the *federation_result* message, I agree, but this is a problem we have to deal with any way looking forward to https://github.com/dedis/popstellar/issues/1802. 

A hacky solution for now could be to extend the *federation_expect* message to also include two *federation_result* messages (both signed by Bob) one for the case of success and one for the case of failure. We then however couldn't use the *reason* field in the message anymore.

Probably a better way to do this is: After the checking the challenge the server informs the organizer of the result, who then in turn generates the needed message on demand and returns it back to the server (who then in turn returns it to the server of the other LAO).

This of course means that more network traffic is being needed and also carries some other implications but it also already goes into the direction of the mechanic we'd have to implement in the Message Exchange protocol anyway.

> If im not wrong, the challenge should be signed by S1’ key

The challenge must be signed by Alice.

> Our purpose is to authenticate the server so it should be signed by the server i suppose

We want to authenticate the remote LAO. This does not necessarily mean the server (or rather it shouldn't mean the server), really we want to authenticate the organizer. If we were to authenticate the server we were also bound to that server for all time to come.

Compare this to authenticating the organizer---say for example server S1 fails (or becomes malicious), this would carry implications regarding the connectivity, but the trust boundaries would still be intact!

**Comment by 1florentin** [comment]
In relation to the first part of the previous comment, I added the JSON schema for the *challenge_request* and created a PR (to @arnauds5's current working branch regarding the Schemas). https://github.com/dedis/popstellar/pull/1804
